sql/stats: TestAverageRefreshTime failed

[(sql/stats).TestAverageRefreshTime failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2643714&tab=buildLog) on [master@100c09f4f6eb3f5b18a67ec4bbfdfe989e0d6ce2](https://github.com/cockroachdb/cockroach/commits/100c09f4f6eb3f5b18a67ec4bbfdfe989e0d6ce2):

```
=== RUN   TestAverageRefreshTime
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestAverageRefreshTime964016156
    test_log_scope.go:74: use -show-logs to present logs inline
    automatic_stats_test.go:291: expected avgRefreshTime 4h0m0s but found 3h53m0s
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestAverageRefreshTime964016156
--- FAIL: TestAverageRefreshTime (45.93s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=4

```
make stressrace TESTS=TestAverageRefreshTime PKG=./pkg/sql/stats TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestAverageRefreshTime.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/stats/automatic_stats_test.go
+++ b/pkg/sql/stats/automatic_stats_test.go
@@ -150,6 +150,11 @@ func TestAverageRefreshTime(t *testing.T) {
 	)
 	refresher := MakeRefresher(st, executor, cache, time.Microsecond /* asOfTime */)
 
+	// curTime is used as the current time throughout the test to ensure that the
+	// calculated average refresh time is consistent even if there are delays due
+	// to running the test under race.
+	curTime := timeutil.Now()
+
 	checkAverageRefreshTime := func(expected time.Duration) error {
 		cache.RefreshTableStats(ctx, tableID)
 		return testutils.SucceedsSoonError(func() error {
@@ -178,14 +183,14 @@ func TestAverageRefreshTime(t *testing.T) {
 			if stat == nil {
 				return fmt.Errorf("no recent automatic statistic found")
 			}
-			if !lessThan && stat.CreatedAt.After(timeutil.Now().Add(-1*expectedAge)) {
+			if !lessThan && stat.CreatedAt.After(curTime.Add(-1*expectedAge)) {
 				return fmt.Errorf("most recent stat is less than %s old. Created at: %s Current time: %s",
-					expectedAge, stat.CreatedAt, timeutil.Now(),
+					expectedAge, stat.CreatedAt, curTime,
 				)
 			}
-			if lessThan && stat.CreatedAt.Before(timeutil.Now().Add(-1*expectedAge)) {
+			if lessThan && stat.CreatedAt.Before(curTime.Add(-1*expectedAge)) {
 				return fmt.Errorf("most recent stat is more than %s old. Created at: %s Current time: %s",
-					expectedAge, stat.CreatedAt, timeutil.Now(),
+					expectedAge, stat.CreatedAt, curTime,
 				)
 			}
 			return nil
@@ -232,7 +237,7 @@ func TestAverageRefreshTime(t *testing.T) {
 				return err
 			}
 			createdAt, err := tree.MakeDTimestamp(
-				timeutil.Now().Add(time.Duration(-1*(i*3+7))*time.Hour), time.Hour,
+				curTime.Add(time.Duration(-1*(i*3+7))*time.Hour), time.Hour,
 			)
 			if err != nil {
 				return err
@@ -257,7 +262,7 @@ func TestAverageRefreshTime(t *testing.T) {
 	}
 
 	// Add some stats on column v in table a with name AutoStatsName, separated
-	// by three hours each, starting 6 hours ago.
+	// by four hours each, starting 6 hours ago.
 	if err := s.DB().Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
 		for i := 0; i < 10; i++ {
 			columnIDsVal := tree.NewDArray(types.Int)
@@ -265,7 +270,7 @@ func TestAverageRefreshTime(t *testing.T) {
 				return err
 			}
 			createdAt, err := tree.MakeDTimestamp(
-				timeutil.Now().Add(time.Duration(-1*(i*4+6))*time.Hour), time.Hour,
+				curTime.Add(time.Duration(-1*(i*4+6))*time.Hour), time.Hour,
 			)
 			if err != nil {
 				return err
@@ -316,7 +321,7 @@ func TestAverageRefreshTime(t *testing.T) {
 				return err
 			}
 			createdAt, err := tree.MakeDTimestamp(
-				timeutil.Now().Add(time.Duration(-1*(i*90+300))*time.Minute), time.Minute,
+				curTime.Add(time.Duration(-1*(i*90+300))*time.Minute), time.Minute,
 			)
 			if err != nil {
 				return err

